### PR TITLE
Integrate misc-tests into wasi-common

### PR DESCRIPTION
**Pull request by kubkon**
This PR builds on #28. It integrates tests from [sunfishcode/misc-tests](https://github.com/sunfishcode/misc-tests) into `wasi-common`. In order to achieve more seamless integration and better control of each tested aspect, I've taken the liberty of modularising [sunfishcode/misc-tests](https://github.com/sunfishcode/misc-tests) into separate testcases. The effect of this work can be seen in my fork [kubkon/misc-tests](https://github.com/kubkon/misc-tests).

Summary of changes:
* add wasmtime as the default runtime for integration tests (code essentially taken from [wasmtime/src/wasmtime.rs](https://github.com/CraneStation/wasmtime/blob/master/src/wasmtime.rs))
* add [kubkon/misc-tests](https://github.com/kubkon/misc-tests) as a submodule (NB it is only required to run the tests)

The workspaces used (per each testcase which is essentially a separate executable Wasm module) is generated on the fly and saved in the host's `temp` dir. In order to avoid any race conditions, etc., each testcase module gets its own workspace dir suffixed with the timestamp of execution. This can potentially generate quite a few directories, albeit in host's `temp` dir, so I'm wondering whether there could be a better way of handling this. Any thoughts on this are more than welcome! :-)

~~EDIT: I was thinking of simply disabling all of the tests on Win, but still I'd expect the `buf_random_buf` testcase to pass. Somehow, this is not the case, so I'll have to investigate. Anyhow, otherwise, feel free to review the changes as I don't expect the bulk to change in the process of me fixing the Win build.~~ OK, I've separated out tests which require preopens from those that don't, and now `buf_random_buf` testcase passes just fine :+1:

### CODE REVIEW

**Review by sunfishcode — COMMENTED** [review]
tests/runtime/wasi.rs appears to be duplicating a lot of code from wasmtime. Is there a way we could factor/organize things to minimize the duplication?

Would it be possible to instead import the wasmtime-wasi crate, and use a [patch](https://doc.rust-lang.org/nightly/edition-guide/rust-2018/cargo-and-crates-io/replacing-dependencies-with-patch.html) to replace its wasi-common repo with the local tree?

Another option would be to integrate misc-tests into the wasmtime repo, rather than here in the wasi-common repo. That would provide clearer layering, however it would mean that we wouldn't get many tests that could test wasi-common in isolation. However, it might not be a bad idea to start building up a testsuite for wasi-common that just works in terms of WASI API calls rather than through full wasm programs.

Thoughts?

**Comment by kubkon** [comment]
> tests/runtime/wasi.rs appears to be duplicating a lot of code from wasmtime. Is there a way we could factor/organize things to minimize the duplication?
> 

I hate code duplication as well, so I'm all for maximising code re-use as much as possible.

> Would it be possible to instead import the wasmtime-wasi crate, and use a [patch](https://doc.rust-lang.org/nightly/edition-guide/rust-2018/cargo-and-crates-io/replacing-dependencies-with-patch.html) to replace its wasi-common repo with the local tree?

That sounds like a great idea! I'll give it a go and see where it takes us. I actually didn't know you can do crazy things like this with Cargo, so thanks for pointing it out! :-)

> 
> Another option would be to integrate misc-tests into the wasmtime repo, rather than here in the wasi-common repo. That would provide clearer layering, however it would mean that we wouldn't get many tests that could test wasi-common in isolation. However, it might not be a bad idea to start building up a testsuite for wasi-common that just works in terms of WASI API calls rather than through full wasm programs.
> 
> Thoughts?

That's a good point. In fact, I'm currently also more inclined to have those tests as part of the `wasi-common` crate. It will definitely make it a lot easier to verify any future contributions if we have tests integrated directly into `wasi-common` and run automatically at each PR, etc. This is not to say we shouldn't have a separate testsuite in `wasmtime` as well.

To sum up, I'll try the patch approach first, and if that fails, I guess we can think on how to integrate `misc-tests` with `wasmtime`. Agreed? :-)

**Comment by kubkon** [comment]
@sunfishcode OK, I've successfully managed to import `wasmtime-wasi` and use a patch to replace its `wasi-common` dependency with local copy, so we're in a much better shape now :-)

File: tests/misc_tests.rs [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+    // Run!
+    runtime::instantiate(&data, bin_name, None)
+}
+
+#[cfg(all(unix))]
+#[test]
+fn sched_yield() -> Result<(), String> {
+    run_test_without_workspace("tests/misc-tests/bin/sched_yield.wasm")
+}
```
Comment (sunfishcode) [inline tests/misc_tests.rs]: What would you think of auto-generating these test functions, similar to what [wasm time does for the spec testsuite](
https://github.com/CraneStation/wasmtime/blob/master/build.rs)? That way, when we add new tests, we don't have to remember to add them here before they actually get run.

**Review by sunfishcode — COMMENTED** [review]
Nice

I like your changes to misc-tests to split it into separate programs. What would you think of moving this into a CraneStation repository, named wasi-misc-tests or something?

Right now, your misc-tests repo contains the precompiled wasm files that it builds, and then you pull them in by git submodule here. I find it a little confusing, because this also pulls in the source for these tests too, however we don't actually build from the source here; we only use the precompiled versions. What would you think about checking the precompiled wasms into the wasi-common repo instead, and avoiding the submodule?

**Comment by kubkon** [comment]
> Nice
> 
> I like your changes to misc-tests to split it into separate programs. What would you think of moving this into a CraneStation repository, named wasi-misc-tests or something?

Great idea! I've now cleaned up my fork of `sunfishcode/misc-tests` and submitted a PR upstream. After you review it and merge it, I guess we could transfer the repo to `CraneStation/misc-tests` and rename it to something like `CraneStation/wasi-misc-tests`. If you'd like I'm happy to do it, would just need admin privs to both `sunfishcode/misc-tests` and `CraneStation` ;-)

> 
> Right now, your misc-tests repo contains the precompiled wasm files that it builds, and then you pull them in by git submodule here. I find it a little confusing, because this also pulls in the source for these tests too, however we don't actually build from the source here; we only use the precompiled versions. What would you think about checking the precompiled wasms into the wasi-common repo instead, and avoiding the submodule?

Yeah, that's a fair point. My thinking here was that it would be nice to preserve some context for the test binaries we use. However, this is actually not needed since, in the end, we will still have access to the sources in a separate repo. So now, I've removed the submodule, and instead, as you suggested, bundled the test binaries directly in `wasi-common` in `wasi-common/misc_testsuite` dir. :+1:

File: tests/misc_tests.rs (hunk 1 above)
Comment (kubkon) [reply tests/misc_tests.rs]: SGTM! And this is exactly what I've done now so have a look :-)

**Comment by sunfishcode** [comment]
Cool. If we later decide we'd like to organize things differently, we can change things of course :-).
